### ISSUE TITLE
Remove logic from setup.py

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8622]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)| @markrwilliams reported|
|-|-|
|Trac ID|trac#8622|
|Type|enhancement|
|Created|2016-07-18 15:40:33Z|

As [8606#3 discussed here], there's non-trivial code that's in `setup.py` (and at this time, `setup3.py`, though that will change when [#8606](https://github.com/twisted/twisted/issues/8606) is closed)

It's difficult to test any logic in `setup.py`. Instead, `setup.py` should be made as simple as possible, and as much code as possible should be moved into `twisted.python.dist` where it can be tested.

adiroiban has [done significant work in this direction](https://gist.github.com/adiroiban/5b461f3ca96828a3036c99964d1d236e).

<details><summary>Searchable metadata</summary>

```
trac-id__8622 8622
type__enhancement enhancement
reporter__markrwilliams markrwilliams
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1468856433702271 1468856433702271
changetime__1471816326394242 1471816326394242
version__None None
owner__adiroiban adiroiban

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban removed owner|
|-|-|

PR at https://github.com/twisted/twisted/pull/467

All previous methods which were just returning a static data structure based on python version were removed

setup.py is now just calling twisted.python.dist.setup() and that is all

In previous version we had both data structured and build logic split between setup.py and dist.py. Now everything was moved to dist.py

As a drive-by I have removed admin/run-python3-tests and admin/run-python3-tests-venv as I don't think they we are using them.

--------

twisted.python.dist3 module was renamed as twisted.python._python3_port as
I don't see why we need to have it as a public API

Also, I don't see why twisted.python.dist is a public API. Does external projects need to know about how Twisted is defining its setup() method?

twisted.python.dist.getVersion and build_scripts_twisted were removed without deprecation as they are no longer used in the Twisted code base.

I did not went for the deprecation dance, since the twisted.python.dist docstring starts with `Don't use this outside of Twisted.` ... so this is like a private API.

I will follow up over the mailing list to let other know about these changes.

Meanwhile please check that the general direction of this ticket is OK.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

The updated tests are tricky since then executed inside or outside tox we will have a different working directory... there is a different path for location where the twisted source code is... so that we can run some setup.py tests

I am testing this as:

```
$ ./bin/trial twisted/python/test/test_setup.py
$ tox -e py27-coverage-posix twisted.python.test.test_setup
```

I went for introducing a TOX_INI_DIR env var and I don't have a better idea, I find if better than start mocking/patching setuptools/distutils

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Hi Adi, please address the comments on the PR, then land it.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban set status to `closed`|
|-|-|

In changeset 41482c4fe5442bcc7b046ab53bf15409f2031abd
```
#!CommitTicketReference repository="" revision="41482c4fe5442bcc7b046ab53bf15409f2031abd"
Merge 8622-setup-coverage: Move all setup.py related code to t.python._setup.


Author: adiroiban
Reviewer: hawkowl
Fixes: #8622
```

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Many thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

In changeset f96bc30f14f528bfd6085ad6db85fc238e0f43ef
```
#!CommitTicketReference repository="" revision="f96bc30f14f528bfd6085ad6db85fc238e0f43ef"
Revert "Merge 8622-setup-coverage: Move all setup.py related code to t.python._setup."

Reopens: #8622

'twist' script was left out when resolving conflicts on an update with trunk.
```

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set status to `reopened`|
|-|-|

The new "twist" script was left out during a merge conflict resolution.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban removed owner<br>@adiroiban set status to `new`|
|-|-|

I am putting this back for review without adding tests for console scripts.

I think there is no point to adding unit tests in t.p.test.test_setup as it will just check the declaration of a list.

I think that console scripts can be unit tested using something like this

```
    from pkg_resources import load_entry_point
    args = ['something']
    main = load_entry_point(
        'twisted', 'console_scripts', 'twist')
```

but then we need to make sure that all tests environments are running pip install before running the tests.

There is ongoing work in removing bin/trial and having all buildbot builders executed with tox ... but we are not there yet.

So, after all the dev/test environment have `pip install -e .` or something similar, we can look at adding tests for the scripts.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)|@rodrigc set owner to @adiroiban|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_10"></a>|@rodrigc commented|
|-|-|

https://github.com/twisted/twisted/pull/516#issuecomment-241281488

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban set status to `closed`|
|-|-|

In changeset 24d4b0d1385560437ab32af2c10129851eca6159
```
#!CommitTicketReference repository="" revision="24d4b0d1385560437ab32af2c10129851eca6159"
Merge 8622-setup-coverage: Re-merge refactoring of setup.py


Author: adiroiban
Reviewer: hawkowl, rodrigc
Fixes: #8622
```
